Enable float only requantization. Part 1.

kimishpatel opened the pull request:
This PR is motivated by two issues it tries to address:
1) relax the constraint on requantization scale (<1). 
2) Unify requantization methodology across pytorch integration of QNNPACK and FBGEMM.

Here we are trying to address the first part for Conv and Linear.
Existing requantization scheme performs scale multiplication entirely in integer arithmetic by extracting mantissa and exponent part of FP scale and processing them. This including appropriate rounding required. The set of instruction, corresponding to this, are specifically tailored for the condition when scale < 1.

Relaxing this constraint requires us to fix that sequence of instruction. In this PR we take a simpler approach of essentially converting Int32 to FP32, apply scale, convert FP32 to Int32 with appropriate rounding, to-nearest-ties-to-even. This is followed by zero point add and clipping. Since in 32-bit ARM nearest-ties-to-even rounding instruction is not available, the sequence is little different. Sequence for both 32-bit and 64-bit are taken from https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/requantization/fp32-neon.c.

Furthermore relaxing the scale constraint and moving towards FP requantization also helps us move towards unifying requantization producer across QNNPACK and FBGEMM.


Summary of the PR:
- requantization params are modified to lift some computation that would have to be in the kernel otherwise for aarch32 kernels, particularly:
   - Computing vfmin, vfmax, vfmagic and vimagic.
- Fixed q8gemm, q8conv and q8dwconv kernels.
- Fixed the corresponding tests.

What is not done:
- XZP kernels are not changed as part of this PR.

kimishpatel committed:
Enable float only requantization. Part 1.

Summary:
- Fix _conv_ params.
- Fixed src/q8gemm/2x4c8-sse2.c, src/q8gemm/4x4c2-sse2.c kernels.
- Fixed src/q8gemm/4x8-neon.c kernel.

Fixed neon as well as asm kernels for q8gemm.
No xzp kernels. Dual issue kernels won't be fixed as part of this.

Changes for requant scale of q8conv
SSE2, neon and asm kernels changed.

Refacatored requant params and updated all arch32 kernels.
Computing vfmin, vfmax, vfmagic and vimagic adds too much overhead.
These values can be computed as params. So extended params structure
to hold these auxiliary values. These are only used in aarch32 kernels.

1. Fixed q8gemm, q8conv arch32 kernels.
2. Fixed q8dwconv kernels.

Modified gemm and convolution tests for requant scale changes.

Some bug fixes.

AshkanAliabadi commented:
Can you please provide some background, preferably as part of the git commit message, to explain not only the what, but also the why of the change?  That would allow me and I think future programmers revisiting this change understand the context better.  Thanks!

kimishpatel commented:
@AshkanAliabadi, update summary.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/requantization.h:
```diff
@@ -55,6 +55,31 @@ pytorch_qnnp_compute_scalar_requantization_params(
+    uint8_t min,
+    uint8_t max) {
+  /* Compute requantization parameters */
+  assert(scale < 1.0f);
```
is this the requantization scale? Why do we have this check here?

kimishpatel replied:
So my plan was to first land changes for doing requantization via FP first. So that is one PR. And then we switch off the scale constraint in a separate PR, so it is easy to revert if necessary.

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/requantization.h:
```diff
@@ -55,6 +55,31 @@ pytorch_qnnp_compute_scalar_requantization_params(
+      (int32_t)(uint32_t)zero_point));
+  params.scalar.max_less_zero_point = ((float)((int32_t)(uint32_t)max -
+      (int32_t)(uint32_t)zero_point));
+  params.scalar.magic = 12582912.0f;
```
Could you add a comment here explaining what these magical variables do?

kimishpatel replied:
I will copy paste the comment from elsewhere, where this is copied from. It is still little convoluted to wrap head around :).

kimishpatel replied:
@supriyar, here is the comment that explains it: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/requantization/fp32-neon.c#L97

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/requantization.h:
```diff
@@ -128,48 +153,23 @@ static inline union pytorch_qnnp_conv_quantization_params
-  const int32_t shift = 127 + 31 - 32 - (fp32_to_bits(scale) >> 23);
-  assert(shift >= 0);
-  assert(shift < 32);
+  assert(scale > 0);
```
Same questions about the scale range here too

kimishpatel replied:
I think this constraint makes sense. I dont think scale is supposed to be < 0, right?

supriyar replied:
Yeah this is ok, I meant the restriction that scale <=1. In the `pytorch_qnnp_compute_scalar_fp32_requantization_params` we have scale < 1.0f but here it is <=1. Why the difference?

kimishpatel replied:
Oh good catch. This should be < 1.0 too although it will be removed so wont really matter.

supriyar reviewed (commented):
Can we have some tests which verify that this is actually in line with fbgemm results? We could have a conv/linear layer that produces requantization scale > 1 and compare result with fbgemm.
cc @dskhudia

kimishpatel commented:
> Can we have some tests which verify that this is actually in line with fbgemm results? We could have a conv/linear layer that produces requantization scale > 1 and compare result with fbgemm.
> cc @dskhudia

Thats a good idea. Let's do that in the PR when we relax the scale constraint. One thing I want to point out though is the FBGEMM implementation here: https://github.com/pytorch/FBGEMM/blob/master/src/QuantUtilsAvx2.cc#L537, has the constraint on the range for FP32 value that we get so conversion to Int32 does not produce -1. I suggest we add clipping there. I need to make similar check for ARM as well and see, what it does for FP values which are larger than INT32_MAX.

supriyar commented:
> > Can we have some tests which verify that this is actually in line with fbgemm results? We could have a conv/linear layer that produces requantization scale > 1 and compare result with fbgemm.
> > cc @dskhudia
> 
> Thats a good idea. Let's do that in the PR when we relax the scale constraint. One thing I want to point out though is the FBGEMM implementation here: https://github.com/pytorch/FBGEMM/blob/master/src/QuantUtilsAvx2.cc#L537, has the constraint on the range for FP32 value that we get so conversion to Int32 does not produce -1. I suggest we add clipping there. I need to make similar check for ARM as well and see, what it does for FP values which are larger than INT32_MAX.

Yes lets do that as well when you update the scale constraint. 
I was mainly concerned about numerics. Even though in this PR you aren't relaxing the scale constraint, the conversion to int8 is still using a different calculation than before. We need to make sure all tests pass and numerics aren't too off. Is there a way to check this in the qnnpack test framework? i.e. previous implementation with scale multiplication in integer arithmetic gives same results as this new one.

I'm going to leave the review of the assembly and intrinsics changes to @AshkanAliabadi.

AshkanAliabadi commented:
Two questions:

1) Do all QNNPACK unit tests pass on all platforms - x86, ARM32, ARM64 both on Android and iOS including NEON and assembly?  I assume yes but considering that they are not run as part of the CI, I wanted to double check.

2) What's the timing like?  There are benchmarks there as well.  How much of a performance penalty are we looking at?

kimishpatel commented:
> Yes lets do that as well when you update the scale constraint.
> I was mainly concerned about numerics. Even though in this PR you aren't relaxing the scale constraint, the conversion to int8 is still using a different calculation than before. 

Do you mean this implementation is doing it differently by going through FP path?

> We need to make sure all tests pass and numerics aren't too off. Is there a way to check this in the qnnpack test framework? i.e. previous implementation with scale multiplication in integer arithmetic gives same results as this new one.
> 
I dont think that is the equivalence test we are looking for. Previous implementation is not the base for accuracy comparison, right? If we are trying to reconcile requantization differences between different frameworks, then that is the test I suggest we add. E.g., A set of fixed inputs run on FBGEMM and QNNPACK (narrow it down to ARM32/64) and require them to produce the same results.

supriyar commented:
> 
> Do you mean this implementation is doing it differently by going through FP path?
> 
Yes
> I dont think that is the equivalence test we are looking for. Previous implementation is not the base for accuracy comparison, right? If we are trying to reconcile requantization differences between different frameworks, then that is the test I suggest we add. E.g., A set of fixed inputs run on FBGEMM and QNNPACK (narrow it down to ARM32/64) and require them to produce the same results.

That's true, but we don't want huge variations with what was there before either. Is it possible to use QNNPACK unit tests for this?

kimishpatel commented:
> That's true, but we don't want huge variations with what was there before either. Is it possible to use QNNPACK unit tests for this?

QNNPACK unit-test are generally use the same requant scheme that kernels use. So as part of this PR I had to change them as well. The unit tests are mostly checking the core algorithm. I can try to revert requant scheme in unit test to older one, but I am wondering if we should do model level tests on top of this PR.

kimishpatel commented:
@AshkanAliabadi 
> Two questions:
> 
> 1. Do all QNNPACK unit tests pass on all platforms - x86, ARM32, ARM64 both on Android and iOS including NEON and assembly?  I assume yes but considering that they are not run as part of the CI, I wanted to double check.

The checks are done for android only , not for ios. I am not sure if there is a straighforward way to run unittest on ios. Is there a reason why you believe that checking on android won't be sufficient, since are only looking at the changes which are only architecture specific.
On android, x86, ARM32/64 pass. XZP kernels have been failing but thats true in the master.

> 2. What's the timing like?  There are benchmarks there as well.  How much of a performance penalty are we looking at?

Generally speaking there is a perf benefit. 
| bench | Relative speedup, arm32  | Relative speedup arm64 |
| ------------- | ------------- |
| Conv  | 4.5%  | 10.9% |
| Linear  | 0.4%  | 5.7% |

Although in Conv, variance is very high
Did not compare for SSE2.

kimishpatel commented:
> That's true, but we don't want huge variations with what was there before either. Is it possible to use QNNPACK unit tests for this?
@supriyar: So I did do this check for the sake of it. That is if the new scheme is compatible with old one or not. And tests fail. I should have realized this earlier, but the old scheme does rounding in two places (this is how gemmlowp does). One is during integer multiply (arm instruction for this is the one that provides rounding). And the second one is happens during shift (there is rounding shift instruction in arm) that accounts for the exponent multiply. Moving to FP requantization now we have only one rounding, so this is not quite compatible as such. Hopefully this helps with one-off differences that we may have between fbgemm and qnnpack.

kimishpatel commented:
@AshkanAliabadi, can you take a look at the PR?

kimishpatel committed:
Removed scale assert.
And fixed the onnx test.

Summary:
Due to change in requant logic now c2's qnnpack and pt integrated
qnnpack results can be off by 1.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar commented on test/onnx/test_pytorch_onnx_caffe2_quantized.py:
```diff
@@ -34,7 +34,13 @@ def generic_test(self, model, sample_inputs, input_names=None, decimal=3):
+        # Permute pytorch output to NHWC
+        # This check had to be changed to account for changes in
+        # qnnpack's requant logic.
+        np.testing.assert_(max_diff <= 1, "Maximum absolute difference must be less than 1")
```
I think we shouldn't be relaxing the check for all tests. Currently the requant changes will affect only conv and linear ops. For other ops we should still match with C2.

kimishpatel replied:
This function is used in multiple places. Maybe we can add a flag for relaxed test. However, the problem is this test is flaky. So imagine you add a new test with conv/linear ops in it and dont set relaxed_compare=True. The test may pass then but it may randomly break. So then you will have this test broken in master. Since this test is data-dependent there is some risk of that. If we are willing to live with it and fix as things break, I am ok with it, but I want to point this out.

supriyar replied:
Yeah I think adding a flag is fine. Currently it only needs to be set for the `test_small_model` case as the conv and linear tests seem to have their own checking. Please add a comment explaining when/why the flag needs to be set.

kimishpatel replied:
@supriyar Done.

supriyar commented on test/onnx/test_pytorch_onnx_caffe2_quantized.py:
```diff
@@ -34,7 +34,13 @@ def generic_test(self, model, sample_inputs, input_names=None, decimal=3):
+        output_diff = np.absolute(np.squeeze(output.detach().numpy()) - caffe_res)
+        max_diff = np.amax(output_diff)
+
+        # Permute pytorch output to NHWC
```
Why is this comment here?

kimishpatel replied:
It was there from where it is copy pasted from. Will remove.

raghuramank100 commented:
Does this change affect the tests here?.
Also, what is the requantization logic for qnnpack on x86, are we changing that also?
https://github.com/pytorch/pytorch/blob/2cf53128a8212b793ee1196fcea15434ce6df602/test/quantization/test_quantized.py#L2030

kimishpatel commented:
> Does this change affect the tests here?.
> Also, what is the requantization logic for qnnpack on x86, are we changing that also?
> https://github.com/pytorch/pytorch/blob/2cf53128a8212b793ee1196fcea15434ce6df602/test/quantization/test_quantized.py#L2030

@raghuramank100, yes it should affect the test there and if those fail they should be caught.
Also, yes. This PR does make changes for x86 as well.

supriyar reviewed (approved):
Looks good, thanks!
